crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+                }
+            }
+        }
+        self.coeffs.push(coeff);
+        self.boundaries.push(self.bit_terms.len());
+        Ok(())
+    }
+}
+
+#[pymethods]
+impl SparseObservable {
+    #[pyo3(signature = (data, /, num_qubits=None))]
+    #[new]
+    fn py_new(data: Bound<PyAny>, num_qubits: Option<u32>) -> PyResult<Self> {
+        let py = data.py();
+        let check_num_qubits = |data: &Bound<PyAny>| -> PyResult<()> {
+            let Some(num_qubits) = num_qubits else {
+                return Ok(());
+            };
+            let other_qubits = data.getattr(intern!(py, "num_qubits"))?.extract::<u32>()?;
+            if num_qubits == other_qubits {
+                return Ok(());
+            }
+            Err(PyValueError::new_err(format!(
+                "explicitly given 'num_qubits' ({num_qubits}) does not match operator ({other_qubits})"
+            )))
+        };
+
+        if data.is_instance(PAULI_TYPE.get_bound(py))? {
+            check_num_qubits(&data)?;
+            return Self::py_from_pauli(data);
+        }
+        if data.is_instance(SPARSE_PAULI_OP_TYPE.get_bound(py))? {
+            check_num_qubits(&data)?;
+            return Self::py_from_sparse_pauli_op(data);
+        }
+        if let Ok(label) = data.extract::<String>() {
+            let num_qubits = num_qubits.unwrap_or(label.len() as u32);
+            if num_qubits as usize != label.len() {
+                return Err(PyValueError::new_err(format!(
+                    "explicitly given 'num_qubits' ({}) does not match label ({})",
+                    num_qubits,
+                    label.len(),
+                )));
+            }
+            return Self::py_from_label(&label).map_err(PyErr::from);
+        }
+        if let Ok(observable) = data.downcast::<Self>() {
+            check_num_qubits(&data)?;
+            return Ok(observable.borrow().clone());
+        }
+        if let Ok(vec) = data.extract() {
+            return Self::py_from_list(vec, num_qubits);
+        }
+        if let Ok(vec) = data.extract() {
+            let Some(num_qubits) = num_qubits else {
+                return Err(PyValueError::new_err(
+                    "if using the sparse-list form, 'num_qubits' must be provided",
+                ));
+            };
+            return Self::py_from_sparse_list(vec, num_qubits).map_err(PyErr::from);
+        }
+        Err(PyTypeError::new_err(format!(
+            "unknown input format for 'SparseObservable': {}",
+            data.get_type().repr()?,
+        )))
+    }
+
+    /// The number of qubits the operator acts on.
+    ///
+    /// This is not inferable from any other shape or values, since identities are not stored
+    /// explicitly.
+    #[getter]
+    #[inline]
+    pub fn num_qubits(&self) -> u32 {
+        self.num_qubits
+    }
+
+    /// The number of terms in the sum this operator is tracking.
+    #[getter]
+    #[inline]
+    pub fn num_ops(&self) -> usize {
```

Comment:
I'm pondering whether we could find a slightly better name, since we don't use abbreviations that much (but `num_operators` is also long to type...). How about `terms`? That would allow a natural distinction between `bit_terms` and `terms` 🙂 
```suggestion
    pub fn num_terms(&self) -> usize {
```

Reply:
I'm happy to switch over to `num_terms` if you think that's better, and not distracting with the single-bit terms.  I personally _think_ about them as terms, I think I just got a bit weird about the naming.

Reply:
Done in 4ca948eaa2.